core/src/main/kotlin/org/evomaster/core/problem/api/service/ApiWsStructureMutator.kt

```diff
@@ -25,56 +26,134 @@ import kotlin.math.min
 /**
  * the abstract structure mutator for API based SUT, such as REST, GraphQL, RPC
  */
-abstract class ApiWsStructureMutator : StructureMutator(){
+abstract class ApiWsStructureMutator : StructureMutator() {
 
     companion object {
         private val log: Logger = LoggerFactory.getLogger(ApiWsStructureMutator::class.java)
     }
 
-
-    @Deprecated("External Actions will be moved into EnterpriseActionGroup")
-    private fun <T : ApiWsIndividual> addInitializingExternalServiceActions(
+    private fun <T : ApiWsIndividual> addExternalServiceActions(
         individual: EvaluatedIndividual<*>,
+        /**
+         * TODO add why
+         */
         mutatedGenes: MutatedGeneSpecification?,
+        /**
+         * [ExternalServiceHandler] is not injected in mutator at the moment, also to validate
+         * it's related to REST only sampler is needed
+         */
         sampler: ApiWsSampler<T>
     ) {
-        // TODO: Incomplete code, under development
 
         if (config.externalServiceIPSelectionStrategy == EMConfig.ExternalServiceIPSelectionStrategy.NONE) {
             return
         }
 
         if (sampler is ResourceSampler) {
-            val ind = individual.individual as? T
+            val ind = individual.individual as? ApiWsIndividual
                 ?: throw IllegalArgumentException("Invalid individual type")
 
-            // TODO: Under review
-            val fw = individual.fitness.getAccessedExternalServiceRequests()
+            val esr = individual.fitness.getViewAccessedExternalServiceRequests()
+            if (esr.isEmpty()) {
+                //nothing to do
+                return
+            }
 
-            val actions = mutableListOf<ExternalServiceAction>().plus(
-                sampler.getExternalService().getExternalServiceActions()
-            )
+            ind.seeMainExecutableActions().forEachIndexed { index, action ->
+
+                val parent = action.parent
+                if (parent !is EnterpriseActionGroup) {
+                    //TODO this should not really happen
+                    val msg = "Action is not inside an EnterpriseActionGroup"
+                    log.error(msg)
+                    throw RuntimeException(msg)
+                }
+
+                // If the action already exists mark it as used
+                esr.flatMap { it.value }.forEach { url ->
+                    val existingActions = parent.getViewOfChildren()
+                        .filterIsInstance<HttpExternalServiceAction>()
+                        .filter { it.request.absoluteURL == url }
+
+                    val actions = sampler.getExternalService()
+                        .getExternalServiceActions()
+                        .filter { it.request.absoluteURL == url }
+
+                    if (existingActions.isEmpty()) {
+                        if (actions.isNotEmpty()) {
```

Comment:
if the external service is visited, but it does not have an instance for it. it might be a bug. maybe have `Lazy.assert` here